File: examples/web/src/shell/libraryPick.test.ts
import { describe, expect, it } from "vitest";
import {
  encodeLibraryPick,
  mergeUnified,
  parentPick,
  parseLibraryPick,
  shareUrl,
} from "./libraryPick";

describe("library pick", () => {
  it("round-trips folder and source", () => {
    const folder = { kind: "folder" as const, mode: "news" as const, folder: "Automotive" };
    const source = {
      kind: "source" as const,
      mode: "news" as const,
      folder: "Automotive",
      sourceId: "moto",
    };
    expect(parseLibraryPick(encodeLibraryPick(folder))).toEqual(folder);
    expect(parseLibraryPick(encodeLibraryPick(source))).toEqual(source);
    expect(parentPick(source)).toEqual(folder);
    expect(parentPick({ kind: "all", mode: "news" })).toBeNull();
  });

  it("merges unified newest first and skips empty share urls", () => {
    const rows = mergeUnified(
      [{ mode: "news", id: "a1", title: "Old", publishedAt: 10, url: "https://n.example/1" }],
      [{ mode: "podcast", id: "p1", title: "Show", publishedAt: 20, url: "https://p.example/rss" }],
      [{ mode: "booru", id: "b1", title: "Board", publishedAt: 0, url: "https://b.example/api" }],
      false,
    );
    expect(rows.map((row) => row.id)).toEqual(["p1", "a1", "b1"]);
    expect(shareUrl(rows[1])).toBe("https://n.example/1");
    expect(
      shareUrl({ mode: "news", id: "x", title: "X", publishedAt: 1, url: "  " }),
    ).toBeUndefined();
  });
});

File: examples/android/app/src/test/java/org/aetherfeed/app/ui/navigation/LibraryPickTest.kt
package org.aetherfeed.app.ui.navigation

import org.aetherfeed.app.domain.Article
import org.aetherfeed.app.domain.Feed
import org.aetherfeed.app.domain.ModuleKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [26])
class LibraryPickTest {
    @Test
    fun encodeRoundTripAndParent() {
        val folder = LibraryPick.Folder(AppDestination.News, "Automotive")
        val source = LibraryPick.Source(AppDestination.News, "Automotive", "moto")
        assertEquals(folder, parseLibraryPick(encodeLibraryPick(folder)))
        assertEquals(source, parseLibraryPick(encodeLibraryPick(source)))
        assertEquals(LibraryPick.Unified, parseLibraryPick("unified"))
        assertEquals(LibraryPick.All(AppDestination.News), folder.parent())
        assertEquals(folder, source.parent())
        assertNull(LibraryPick.All(AppDestination.News).parent())
        assertEquals(3, modeTabDestinations().size)
        assertTrue(AppDestination.entries.none { it.name == "Unified" })
    }

    @Test
    fun sortAndMergeUnified() {
        val news = articlesToItems(
            listOf(Article("a1", "f1", "Old", "https://n.example/1", publishedAt = 10)),
        )
        val pods = feedsToItems(
            listOf(Feed("p1", "Show", "https://p.example/rss", ModuleKind.Podcast, updatedAt = 20)),
            AppDestination.Podcasts,
            ModuleKind.Podcast,
        )
        val boards = feedsToItems(
            listOf(Feed("b1", "Board", "https://b.example/api", ModuleKind.Booru, updatedAt = 0)),
            AppDestination.Booru,
            ModuleKind.Booru,
        )
        val newest = mergeUnified(news, pods, boards, oldestFirst = false)
        assertEquals(listOf("p1", "a1", "b1"), newest.map { it.id })
        assertEquals("https://n.example/1", shareUrl(newest.first { it.id == "a1" }))
        assertNull(shareUrl(LibraryItem(AppDestination.News, "x", "X", 1, "  ")))
        assertTrue(mergeUnified(emptyList(), emptyList(), emptyList(), false).isEmpty())
    }

    @Test
    fun folderPickScopesNewsFeedsAndLastGenWins() {
        val auto = Feed("m", "Moto", "https://m.example/rss", ModuleKind.News, updatedAt = 1, folder = "Automotive")
        val other = Feed("z", "Other", "https://z.example/rss", ModuleKind.News, updatedAt = 1, folder = "Games")
        val scoped = org.aetherfeed.app.news.newsFeedsForPick(
            LibraryPick.Folder(AppDestination.News, "Automotive"),
            listOf(auto, other),
            { feed -> feed.folder.orEmpty() },
        )
        assertEquals(listOf("m"), scoped.map { it.id })
        assertTrue(org.aetherfeed.app.ui.news.acceptPickGen(2, 2))
        assertTrue(!org.aetherfeed.app.ui.news.acceptPickGen(1, 2))
    }
}
